File: ide-common/src/main/java/org/digma/intellij/plugin/psi/CanInstrumentMethodResult.java
package org.digma.intellij.plugin.psi;


public class CanInstrumentMethodResult{

    private final IFailureCause failureCause;

    public CanInstrumentMethodResult(){
        this.failureCause = null;
    }

    public CanInstrumentMethodResult(IFailureCause failureCause){

        this.failureCause = failureCause;
    }

    public static CanInstrumentMethodResult Failure(){
        return new CanInstrumentMethodResult(new GenericFailureCause());
    }

    public boolean wasSucceeded() { return failureCause == null; }

    public IFailureCause getFailureCause(){ return failureCause; }


    public interface IFailureCause {}

    public record MissingDependencyCause(String dependency) implements IFailureCause {}

    public record GenericFailureCause() implements IFailureCause {}
}

File: ide-common/src/main/kotlin/org/digma/intellij/plugin/instrumentation/CanInstrumentMethodResult.kt
package org.digma.intellij.plugin.instrumentation


open class CanInstrumentMethodResult(val failureCause: IFailureCause? = null) {


    companion object {
        @JvmStatic
        fun failure(): CanInstrumentMethodResult {
            return CanInstrumentMethodResult(GenericFailureCause())
        }
    }


    fun wasSucceeded(): Boolean {
        return failureCause == null
    }

//    fun getFailureCause(): IFailureCause? {
//        return failureCause
//    }


}


interface IFailureCause

class MissingDependencyCause(val dependency: String) : IFailureCause

class GenericFailureCause : IFailureCause